DataSetExplorer/DataSetBuilder/Model/Repository/DataSetDatabaseRepository.cs

```diff
@@ -59,5 +59,50 @@ public void Update(DataSet dataSet)
             }
             return result;
         }
+
+        public DataSet DeleteDataSet(int id)
+        { 
+            var deletedDataset = _dbContext.DataSets.Remove(FindDataSetForDeletion(id)).Entity;
+            _dbContext.SaveChanges();
+            return deletedDataset;
+        }
+
+        private DataSet FindDataSetForDeletion(int id)
+        {
+            return _dbContext.DataSets
+                .Include(d => d.SupportedCodeSmells)
+                .Include(s => s.Projects).ThenInclude(p => p.CandidateInstances).ThenInclude(c => c.Instances).ThenInclude(i => i.Annotations).ThenInclude(a => a.ApplicableHeuristics)
+                .Include(s => s.Projects).ThenInclude(p => p.CandidateInstances).ThenInclude(c => c.Instances).ThenInclude(i => i.Annotations).ThenInclude(a => a.InstanceSmell)
+                .FirstOrDefault(s => s.Id == id);
+        }
+
+        public DataSet UpdateDataSet(DataSet dataset)
+        {
+            var updatedDataset = _dbContext.Update(dataset).Entity;
+            _dbContext.SaveChanges();
+            return updatedDataset;
+        }
+
+        public DataSetProject DeleteDataSetProject(int id)
+        {
+            var deletedProject = _dbContext.DataSetProjects.Remove(FindProjectForDeletion(id)).Entity;
+            _dbContext.SaveChanges();
+            return deletedProject;
+        }
+
+        private DataSetProject FindProjectForDeletion(int id)
+        {
+            return _dbContext.DataSetProjects
+                .Include(p => p.CandidateInstances).ThenInclude(c => c.Instances).ThenInclude(i => i.Annotations).ThenInclude(a => a.ApplicableHeuristics)
```

Comment:
Not sure how EF handles this, but do you need all these includes if you configured cascade deletion below?

Reply:
You are right, i don't need this. I added some things that were missing in the cascade delete configuration.